kotlin/workflow-testing/src/main/java/com/squareup/workflow/testing/RenderTester.kt

```diff
@@ -0,0 +1,107 @@
+ *   // Test rendering from initial state…
+ *   .render(
+ *      props = MyProps(…),
+ *      expectations = RenderExpectations(
+ *        expectedWorkers = listOf(
+ *          ExpectedWorker(
+ *            matchesWhen = { it is SubmitLoginWorker },
+ *            key = "signin",
+ *            output = EmittedOutput(LoginResponse(success = true))
+ *          )
+ *        ),
+ *        expectedWorkflows = listOf(
+ *          ExpectedWorkflow(
+ *            workflowType = ChildWorkflow::class,
+ *            key = "child",
+ *            assertProps = { assertThat(it.email).isEqualTo("test@foo.com") },
+ *            rendering = ChildRendering("message")
+ *          )
+ *        ),
+ *        assertState = { assertTrue(it is FooState) },
+ *        assertOutput = { assertTrue(it is LoginSuccessful) },
+ *        assertRendering = { assertTrue(it is LoadingScreen) }
+ *      )
+ *   )
+ *   // State has now transitioned, test rendering with the new state…
+ *   .render(
+ *     …
+ *   )
+ * ```
+ *
+ * It is an error if more than one workflow or worker specifies an output.
+ */
+fun <PropsT, OutputT : Any, RenderingT> Workflow<PropsT, OutputT, RenderingT>.renderTester()
+    : RenderTester<PropsT, Nothing, OutputT, RenderingT> {
+  @Suppress("UNCHECKED_CAST")
+  return (asStatefulWorkflow() as StatefulWorkflow<PropsT, Unit, OutputT, RenderingT>)
+      .renderTester(Unit) as RenderTester<PropsT, Nothing, OutputT, RenderingT>
+}
+
+fun <PropsT, StateT, OutputT : Any, RenderingT>
+    StatefulWorkflow<PropsT, StateT, OutputT, RenderingT>.renderTester(
+  initialState: StateT
+): RenderTester<PropsT, StateT, OutputT, RenderingT> = RealRenderTester(this, initialState)
+
+interface RenderTester<PropsT, StateT, OutputT, RenderingT> {
+  fun render(
+    props: PropsT,
+    expectations: RenderExpectations<StateT, OutputT, RenderingT>
+  ): RenderTester<PropsT, StateT, OutputT, RenderingT>
+}
+
+data class RenderExpectations<StateT, OutputT, RenderingT>(
+  val expectedWorkers: List<ExpectedWorker<*>> = emptyList(),
+  val expectedWorkflows: List<ExpectedWorkflow<*, *, *>> = emptyList(),
+  val assertState: (state: StateT) -> Unit = {},
+  val assertOutput: (output: OutputT) -> Unit = {},
+  val assertRendering: (rendering: RenderingT) -> Unit
+)
+
+/**
+ * @param matchesWhen Predicate used to determine if this matches the worker being ran.
+ * @param output If non-null, [EmittedOutput.output] will be emitted when this worker is ran.
+ */
+data class ExpectedWorker<OutputT>(
+  val matchesWhen: (otherWorker: Worker<*>) -> Boolean,
+  val key: String = "",
+  val output: EmittedOutput<OutputT>? = null
+)
+
+/**
+ * @param assertProps Assertions on the props passed to `renderChild`.
+ * @param rendering The rendering to return from `renderChild` when this workflow is rendered.
+ * @param output If non-null, [EmittedOutput.output] will be emitted when this worker is ran.
+ */
+data class ExpectedWorkflow<PropsT, OutputT : Any, RenderingT>(
```

Comment:
Would it be possible to pass workflow type here? `workflowType` would become unnecessary.

Reply:
You can't only specify a subset of type parameters, so you'd have to also explicitly pass every other type. It would be gross.